### ISSUE TITLE
Export pipeline config to pipeline-as-code format

### ISSUE THREAD

**Issue by arvindsv** [gocd/gocd#5338]
Epic: #5175 (Thoughts about improving pipelines as code (config-repo))

##### Summary

The ability to export pipeline configs, defined in the GoCD config XML, into formats such as JSON/YAML which can then be checked into [pipelines-as-code](https://docs.gocd.org/current/advanced_usage/pipelines_as_code.html) repositories.

##### Details:

See PR: https://github.com/gocd/gocd/pull/5330

**Comment by arvindsv** [comment]
@ibnc I've been thinking about where this feature should be exposed and I think it should be part of the actions in /go/admin/pipelines. I think it should look like this:

<img width="1323" alt="2018_10_26_11-41-29" src="https://user-images.githubusercontent.com/172235/47577210-5006db00-d914-11e8-880f-de3f70f3279d.png">

<img width="1321" alt="2018_10_26_11-41-14" src="https://user-images.githubusercontent.com/172235/47577220-572de900-d914-11e8-9dad-92350ec6401e.png">

When clicked, for now, I'd expect to be shown the JSON/YAML in a way that makes it easy to copy/save.

Let me know what you think.

Of course, I'd expect this to be backed by a proper, documented API call, which allows you to export it as a YAML or JSON. That's what you'd use, if you were exporting multiple pipelines.

**Comment by ibnc** [comment]
Do we only want to allow admins to export pipelines?

**Comment by arvindsv** [comment]
@ibnc Yes, I think so. Currently, we don't let view/operate users view the configuration. The response, in the case of the export, can contain encrypted secrets, which might not be a good idea to show (even though they're encrypted).

**Comment by ibnc** [comment]
PRs have been opened for the json/yaml plugins. The yaml PR is still in progress atm

JSON: https://github.com/tomzo/gocd-json-config-plugin/pull/19
YAML: https://github.com/tomzo/gocd-yaml-config-plugin/pull/86

**Comment by rajiesh** [comment]
@arvindsv some thoughts regarding the export feature,

- While I agree that `Export as` option should be provided in the `/go/admin/pipelines` page, I feel that the `Export as` option should be a Pipelines level option and not at individual pipeline. So that user can  select multiple pipeline and then choose to export. Why I think so is, as a user, I would use `Export as` option to move my pipelines from config xml to config repo. Pipelines will mostly have dependencies and config xml pipeline cannot be a downstream of config repo pipeline, so all dependent pipelines too need to be exported. So `Export as` option at individual pipeline level could be something good to try out the feature but in real time I would prefer it to be a top level option. And I don't see any reason to provide it at both places

- Consider the pipeline to be exported is part of an environment, the sequence of steps user have to follow to export them are,

     - Export the pipeline
     - Remove the pipeline from environment
     - Delete the pipeline from config XML
     - Add the pipeline to config repo and to the environment

Can these steps be simplified? While exporting if the pipeline is part of environment can environments JSON/YAML too be generated

**Comment by zerolaser** [comment]
@arvindsv Once the pipeline is converted as code. The gearbox for that pipeline doesn't work any more. I understand it was meant to do that. but GoCD should still be able to show what each stage is doing and what the config of the pipeline looks like. I highly recommend you to implement this. 
when you click on the gear box it should show all the stages,jobs and tasks like how it normally shows, since it is written as pipeline as code, just give the read only permissions to view here.

**Comment by tomzo** [comment]
I'd like to add to @rajiesh comments, that if we are going to allow exports of individual pipelines, then we should enable it only on a pipeline which is not a dependency of something defined in XML.

**Comment by arvindsv** [comment]
@rajiesh  Yes, what you say makes sense. We need to think more about the interaction. Especially, the part where you list the steps they need to take to move pipelines over to repos.

/cc @tinozza @naveenbhaskar - We need to talk about this.

@tomzo I see what you mean. If you export one pipeline and it is a dependency of some other pipeline which is in the XML, then either both pipelines should be exported or neither. If we add export at a top level, as @rajiesh suggests, then this can be done. It's more effort though.

My other consideration is: Export is a one-time activity for most users. How much effort do we put into it, especially since this is going to be used by slighly-more advanced users, possibly? I'm not saying we shouldn't. Just weighing it.

**Comment by tomzo** [comment]
> My other consideration is: Export is a one-time activity for most users. 

Only at first glance. I think I could easily recommend a total newbie to GoCD to create a pipeline with UI, go through entire wizard, make sure it is working as expected and then, if he/she wants to use config-repo plugin, to export the pipeline and commit it. 
So that would be a one more use case for non-advanced users.

@arvindsv also consider that if this is intended for one-time activity of moving to config-repos, then what is the actual utility of exporting pipelines one by one, if user has to go though a few steps for each of the possibly ~100 pipelines?

**Comment by arvindsv** [comment]
> @arvindsv also consider that if this is intended for one-time activity of moving to config-repos, then what is the actual utility of exporting pipelines one by one, if user has to go though a few steps for each of the possibly ~100 pipelines?

I would focus more on the API than exporting them one by one. The API is what anyone (well, I) would be using if I wanted to export 100s of pipelines. To me, that's most important.

The UI will make it easier, for sure. However, if it's not a regular activity, I'm thinking about how much effort it will be and whether it is worth it to spend time on it, rather than on other things such as improved authorization, secrets management, etc.

**Comment by arvindsv** [comment]
Oh, I agree, based on what @rajiesh said, that there isn't much utility in exporting pipelines one by one.

**Comment by ibnc** [comment]
I also agree with @rajiesh on the one by one exports not being super useful. With that being said, I think it's a good first step to implement before we start considering an "export all the things" type button. 

I think the best order for implementation might be something like this:

1. One by one export (WIP)
2. Export multiple pipelines
3. Export referenced templates with the pipelines being exported
4. Export an Environment (pipelines and all)
5. Export multiple Environments
6. ???

**Comment by ibnc** [comment]
> @arvindsv Once the pipeline is converted as code. The gearbox for that pipeline doesn't work any more. I understand it was meant to do that. but GoCD should still be able to show what each stage is doing and what the config of the pipeline looks like. I highly recommend you to implement this.
> when you click on the gear box it should show all the stages,jobs and tasks like how it normally shows, since it is written as pipeline as code, just give the read only permissions to view here.

@zerolaser  I think what your saying is a little outside the scope of this issue (more so related to https://github.com/gocd/gocd/issues/5175 IMO). Regardless, I completely agree with you. There've been multiple occasions where I've wanted to see the config for a repo configured pipeline in the UI. It'd be much easier than having to go look at the code

**Comment by kierarad** [comment]
> I also agree with @rajiesh on the one by one exports not being super useful. With that being said, I think it's a good first step to implement before we start considering an "export all the things" type button.
> 
> I think the best order for implementation might be something like this:
> 
> 1. One by one export (WIP)
> 2. Export multiple pipelines
> 3. Export referenced templates with the pipelines being exported
> 4. Export an Environment (pipelines and all)
> 5. Export multiple Environments
> 6. ???
> 
> @arvindsv @tomzo @rajiesh @marques-work @kierarad thoughts?

I think most of my thoughts have been captured throughout various comments. The only thing that I would reiterate is the importance of being able to select which pipelines you'd like to export when exporting more than one. I say this because I would assume that if someone is moving their pipeline configuration into config repos there is a good chance they would want to move some pipelines into one repo and other pipelines into a different repo. (Basically I'd advise against an export all sort of mechanism)

Also as part of exporting multiple pipelines/environments I imagine we'd need to limit the number of pipelines exported at any given time?

**Comment by tomzo** [comment]
@ibnc I spoke multiple times about this with @arvindsv and @ketan before, please see also part of discussion in here: 3 comments after https://github.com/gocd/gocd/issues/5175#issuecomment-423830739 and I think it would be best if you opened a new issue for this proposal.

**Comment by ibnc** [comment]
Along the same lines as what @kierarad brought up, we need to think about what we give back to the user when they export more than one pipeline. Go is agnostic to the format that it'll receive from the plugin. For example, the yaml plugin can shove everything into 1 file, but the json plugin requires each pipeline be in their out file.

**Comment by arvindsv** [comment]
@ibnc Let's go with "One by one export" at this time. We will need to design the bulk export better. I'd like to see how people use the API first.

**Comment by ibnc** [comment]
What do we want the format of the export to be for users?

* Should a modal pop up with the plain text in it?
* Should it start a file download?
* ?

**Comment by arvindsv** [comment]
@ibnc I'd expect it to show up in a modal. _If_ we're going to redo it, to handle multiple pipelines, etc. I'd worry about the UI then.

**Comment by ibnc** [comment]
@arvindsv That's what I was thinking.

I think that if we go for multiple pipelines, then we could have it be a tarball download in the future.

**Comment by tomzo** [comment]
IMO in this iteration we should stay at exporting just one pipeline and show a modal which only wraps the config lines returned by plugin.
In longer term, it is the plugin that decides how to structure the output file(s) and I think server should delegate this decision to the plugin. 
Plugins already are aware of multiple files for `parse-directory` command, so theoretically the reverse operation should return a directory of files. So yes, tarball does make sense.

**Comment by arvindsv** [comment]
@ibnc @kierarad About this:

> > What do we want the format of the export to be for users?

> >  -  Should a modal pop up with the plain text in it?
> >  -  Should it start a file download?

I'd said:
> I'd expect it to show up in a modal. If we're going to redo it, to handle multiple pipelines, etc. I'd worry about the UI then.

Was talking to @ketan and he felt that doing a file download would be easier and seeing a big JSON/YAML in a modal is not ideal anyway. I think that makes sense. Can we make it do a file download instead?

**Comment by rajiesh** [comment]
Closing this issue since we are done with first iteration of export pipelines feature with ability to export one pipeline at a time. Other points discussed like multiple pipeline export, simplify UX for moving pipeline to pipeline-as-code(in case of pipelines with dependencies, pipelines in environment) can be done with future iterations
